### PR TITLE
[ENH] Minigallery can take arbitrary files/glob patterns as input

### LINKED ISSUE #1029: Create mini-gallery with an arbitrary list of gallery items
Hi there :slightly_smiling_face: 

For now, mini-galleries filter out gallery items based on whether they use a specified method / class / etc or not.
I think it'd be very nice to be able to generate mini-galleries from an arbitrary list of items specified by the user.
Specifically, I'm planning to use this in `nilearn`'s user guide to reference examples using the nice `sphinx-gallery` thumbnail.

It could look something like this:
```
.. minigallery::
    :add-heading: A list of examples relevant to the section above
    :sg-items: section_1/item_3.py section_4/item_2.py
```

Let me know what you think!

### PR DESCRIPTION
**Pull request by story645**
Closes #1029

- [x] still have to finish up tests and docs, but wanted some feedback if this was in the right direction.

- [x] If so, I'm also wondering if I could strip the thumbnail parent div out of the backref files so that I could do it in the loop.
- [x] get .py files in .rst folders working - seems to work by default and issue was sorting
- [x] file sort order

### CODE REVIEW

**Comment by larsoner** [comment]
> If so, I'm also wondering if I could strip the thumbnail parent div out of the backref files so that I could do it in the loop.

Not sure if this will break anything or not. But you could do it and we could check to see if things still render properly.

> And if it makes sense to add a ... arg that will look in the folder for a readme that defines the sort order?

I'm not sure how it would work to use RST to define the order, but agree it makes sense to define it somewhere somehow.

A solution that would be more consistent with other order setting mechanics would be to provide a `sphinx_gallery_conf["minigallery_sort_order"]` that's a `callable` that takes whatever it needs to take (probably the current source file and list of examples or something?) and returns the sorted order or so.

**Comment by lucyleeow** [comment]
> I have gone down way too deep an html/xml rabbit hole

I have nothing to contribute here but I just thank you for your hard work.

Also the CIs are failing because scipy docs site certificate has expired https://github.com/scipy/docs.scipy.org/issues/80

**Comment by larsoner** [comment]
> does it matter if the heading is before or in the thumbnail parent div? my bias is for in b/c then it stays together as part of the DOM

I am not really sure but if I had to guess I would say no it shouldn't matter, so feel free to try moving it

**Comment by larsoner** [comment]
> I think a standalone/separate PR would be to make something like a FileOrder(ExplicitOrder) object that can be configured with the file it should look for the ordering

Thinking about this more... I'm not sure we can/should implement this at the SG end, at least to start. You could already implement and use this in `matplotlib` because you can define your own sortkey if you want. For example following this code:

https://sphinx-gallery.github.io/stable/_modules/sphinx_gallery/sorting.html#ExampleTitleSortKey

I'd expect something like the following, where the `...` is whatever you need to do to parse the RST in whatever way matplotlib wants to specify the information:

```
class RSTSortKey:

    def __init__(self, src_dir):
        self.src_dir = src_dir
        readme_txt = Path(src_dir) / "README.rst").read_text("utf-8")
        self.fname_list_orderered = ...  # something you pull from the readme

    def __call__(self, filename):
        return self.fname_list_ordered.index(filename)
```

**Comment by larsoner** [comment]
... in other words, I think the hard/interesting work here is in the `...` above, and that's going to be a bit specific to how matplotlib does things. But assuming that works, it would be great to link from the SG config docs to the matplotlib code showing how the `SortKey`-based framework is general enough to support this use case!

**Comment by larsoner** [comment]
> Is it okay to just apply it to the files? I can't tell how the sorting mechanism is handled for the backreferences and I'm concerned that implementing it would require parsing each backreferences file and reordering and I think that's out of scope for this PR.

Can you clarify what you mean? I don't see why this PR would need to mess with the backreferences files

File: doc/configuration.rst
Comment (larsoner) [inline doc/configuration.rst:625]: Instead of a trivial example here, could you make one that actually does something interesting related to the example above with `numpy.exp ../examples/...`? I think it will be easier for people to understand the use case

File: doc/configuration.rst
Comment (larsoner) [inline doc/configuration.rst:625]: Wouldn't this be better, i.e., sort first based on "plot_4b"-ness then on the filename itself?
```suggestion
    'minigallery_sort_order' : lambda path: (not path.startswith('plot_4b'), path)
```

File: sphinx_gallery/directives.py
Comment (larsoner) [inline sphinx_gallery/directives.py:27]: Think this is required for RST to render properly
```suggestion
    The required argument is a one or more of the following:

```
and https://output.circle-artifacts.com/output/job/ff1728af-0451-4fef-bab5-967229bb60df/artifacts/0/html/gen_modules/sphinx_gallery.directives.html#sphinx_gallery.directives.MiniGallery seems to suggest this as well

**Review by larsoner — COMMENTED** [review]
There are now a lot of empty sections like

https://output.circle-artifacts.com/output/job/ff1728af-0451-4fef-bab5-967229bb60df/artifacts/0/html/gen_modules/sphinx_gallery.directives.html#examples-using-sphinx-gallery-directives-visit-imgsg-latex

I don't see them on `dev`

https://sphinx-gallery.github.io/dev/gen_modules/sphinx_gallery.directives.html#sphinx_gallery.directives.MiniGallery

File: doc/configuration.rst
Comment (larsoner) [inline doc/configuration.rst:629]: Okay one last very annoying (and tricky!) thing -- it's important that whatever is in the `sphinx_gallery_conf` have a stable `__repr__` from run to run, otherwise rebuilds will happen for all files on sequential calls to `sphinx_build` even if no files have changed. Unfortunately, we can't guarantee that `lambda`'s `__repr__` will be the same between subsequent runs of Sphinx. So really this needs to be something like:
```
class MinigallerySortOrder:

    def __repr__(self):
         return "MinigallerySortOrder"

    def __call__(self, fname):
        return ...
```
then you pass `"minigallery_sort_order": MinigallerySortOrder(),` as the conf entry. See the `warning` in https://sphinx-gallery.github.io/stable/configuration.html#sorting-gallery-subsections for example

**Review by larsoner — COMMENTED** [review]
Code looks good to me. Does this get what you need for `matplotlib`?

**Comment by larsoner** [comment]
Okay, let me know once the last doc update is done and I'll merge!

**Review by lucyleeow — COMMENTED** [review]
Just a flyby review and a lot of things are nitpicks.

I noted in tinybuild that the thumbnails are one per line:

image

Would it be possible to make it a few per line, like backreferences does?

File: doc/configuration.rst
Comment (lucyleeow) [reply doc/configuration.rst:627]: Oh maybe an explanation about what we are doing (i'm probably just bad with `key`) ?

**Comment by lucyleeow** [comment]
Ahh it could be me, let me try again!

File: doc/configuration.rst
Comment (lucyleeow) [reply doc/configuration.rst:605]: Maybe this could be clearer?

File: sphinx_gallery/tests/test_full.py
Comment (lucyleeow) [reply sphinx_gallery/tests/test_full.py:1178]: Ahh no worries, lets keep!

**Comment by lucyleeow** [comment]
> let me try again!

Hmm I got the same result... This is the html

```html
</div><p>Test 2-F-G</p>
<div class="sphx-glr-thumbnails"><div class="sphx-glr-thumbcontainer" tooltip="This example tests that the alt text is generated correctly for matplotlib figures."><img alt="" src="_images/sphx_glr_plot_matplotlib_alt_thumb.png" />
<p><a class="reference internal" href="auto_examples/plot_matplotlib_alt.html#sphx-glr-auto-examples-plot-matplotlib-alt-py"><span class="std std-ref">Matplotlib alt text</span></a></p>
  <div class="sphx-glr-thumbnail-title">Matplotlib alt text</div>
</div><div class="sphx-glr-thumbcontainer" tooltip="Setting the Matplotlib backend"><img alt="" src="_images/sphx_glr_plot_matplotlib_backend_thumb.png" />
<p><a class="reference internal" href="auto_examples/plot_matplotlib_backend.html#sphx-glr-auto-examples-plot-matplotlib-backend-py"><span class="std std-ref">Setting the Matplotlib backend</span></a></p>
  <div class="sphx-glr-thumbnail-title">Setting the Matplotlib backend</div>
```

What's the difference between mine and yours?

**Comment by lucyleeow** [comment]
So in mine, I didn't have the last line:

```html
</div></div>
```

it looks like

```html
<p><a class="reference internal" href="auto_examples/plot_matplotlib_backend.html#sphx-glr-auto-examples-plot-matplotlib-backend-py"><span class="std std-ref">Setting the Matplotlib backend</span></a></p>
  <div class="sphx-glr-thumbnail-title">Setting the Matplotlib backend</div>
</div><p>Test 3-F-G-B</p>
```

Which only shows where the difference comes from. I just pulled your branch down so it's up to date :shrug:

Could the difference (between this and other mini-galleries) be because we're using the default `is_backref=False` in `_thumbnail_div` ?

Maybe if it works @larsoner, I'll just put it down to me?

File: doc/configuration.rst
Comment (lucyleeow) [reply doc/configuration.rst]: > Also, should I just keep it simple and only sort on paths?

Sorry I don't follow. As in only provide `path` and not `obj`?

**Comment by lucyleeow** [comment]
Okay false alarm, it was totally my environment, not sure how I managed to stuff it up. Works fine in new clean env.

File: doc/configuration.rst
Comment (lucyleeow) [inline doc/configuration.rst:609]: nit - line length?

File: doc/configuration.rst
Comment (lucyleeow) [inline doc/configuration.rst:632]:
```suggestion
* path: backreferences ``{input}.example`` or ``{gallery_example}.py`` file string
```

File: doc/configuration.rst
Comment (lucyleeow) [inline doc/configuration.rst:628]: Can we make explicit that all minigalleries created with the `minigallery` directive will use the single `minigallery_sort_order` config value?

File: doc/configuration.rst
Comment (lucyleeow) [inline doc/configuration.rst:634]:
```suggestion
The sorting key object passed to :py:func:`sorted` must have the following:
```

File: doc/configuration.rst
Comment (lucyleeow) [inline doc/configuration.rst:610]:
```suggestion
tutorials. The mini-gallery directive therefore supports passing in
a list (space separated) of any of the following:
```

File: doc/configuration.rst
Comment (lucyleeow) [inline doc/configuration.rst:655]:
```suggestion
Sorting is not supported for a set/subset of thumbnails generated from
 :ref:`backreferences <references_to_examples>` i.e., the thumbnails
 linked to a qualified object name input.
```

I get what this means now. It's hard to document this clearly, I tried my best.

**Review by lucyleeow — COMMENTED** [review]
More nits!

File: doc/configuration.rst
Comment (lucyleeow) [reply doc/configuration.rst]: Ohh sorry I was not clear and got mixed up. I did not mean to include backreferences. I mean to make it clear that the same `minigallery_sort_order` is used for all file minigalleries , but now that I think about it I don't think you'd ever want use different sort keys for different galleries. Resolving.

File: doc/configuration.rst
Comment (lucyleeow) [inline doc/configuration.rst:644]: "a gallery of thumbnails" ? or "generates gallery thumbnails"

**Comment by lucyleeow** [comment]
> change minigallery to only sort on path?

I don't use SG to really have an opinion on this. I guess maybe you may want to put backreference thumbnails at the end or start ?

> expand sorting to backreferences? 

I would go with YAGNI for this one.

**Review by larsoner — APPROVED** [review]
Agreed the list is nicer!

@lucyleeow feel free to merge if you're happy

**Comment by lucyleeow** [comment]
Sorry this feel off my radar somehow. I will take a look this weekend or early next week. Sorry for holding this up.

**Review by lucyleeow — APPROVED** [review]
Thank you for all your work. Just some doc nitpicks from me!

File: sphinx_gallery/directives.py
Comment (lucyleeow) [reply sphinx_gallery/directives.py:97]: If that's not too much to ask, thank you!

File: doc/configuration.rst
Comment (lucyleeow) [reply doc/configuration.rst:654]: Ah sorry, I didn't see it in context. Ignore in that case, sorry!

File: doc/configuration.rst
Comment (lucyleeow) [inline doc/configuration.rst:663]: I don't quite follow this. I thought the `sorted` function got the `minigallery_sort_order` function?

**Review by lucyleeow — COMMENTED** [review]
Sorry, another question!

Also could we add a 'minigallery' / `minigallery_sort_order` directive item [hereish](https://github.com/sphinx-gallery/sphinx-gallery/blob/1bdb4b9a4d2000e025c5acd7e1771239d582ac6b/doc/configuration.rst#L53)

**Comment by lucyleeow** [comment]
Thank you! I have added a section on creating your own sort key function, I thought it fit better.

I just noticed that it seems that what is passed to the sort key is the full path to the backreference 'examples' files e.g.,:
`/home/lucy/Documents/sphinx-gallery/doc/gen_modules/backreferences/numpy.exp.examples`
and some weird amalgamation of full and relative path to the example files:
`/home/lucy/Documents/sphinx-gallery/doc/../examples/plot_0_sin.py` ?

If you can confirm on your system, maybe we could fix and amend the test to catch this?

**Comment by lucyleeow** [comment]
> So pushed into a separate commit in case you want to reject, because I understand why you'd want to frontload the FunctionSortKey but I pulled the repr discussion up 'cause I thought the section flow was jumpy otherwise.

That's fine, I've used this ordering.

Some minor changes:

* I've just used `.name` to get the filename instead of the full path (it's not pretty but we were manipulating the sort elements anyway so :shrug:  - did not find a way to test this directly but I have amended `MockSort` so it will fail if not the filename
* I've allowed `None` as a `minigallery_sort_order` input - could not think of way to test this directly, though i noticed we don't test this for `subsection_order` either
* I don't think the doc config `minigallery_sort_order` in the SG docs was working (as we were passing the full path before) and I didn't think it was necessary. Also now something fails if `minigallery_sort_order` does not allow `None`
* Doc amendments. I have realised that you cannot use `FunctionSortKey` for `within_subsection_order` config as they get instantiated with the source dir

**Comment by lucyleeow** [comment]
Thanks @story645 for all your work. Just made minor fixes to save you time.

**Comment by lucyleeow** [comment]
Ah sorry I missed that. What making the making the paths relative to `src_dir` ?

We could probably do that here:
```python
            elif paths := Path(src_dir).glob(obj):
                file_paths.extend([(obj, p) for p in paths])
```

?

Maybe we could add a test to capture this? And update the documentation so people know what we're sorting on.

Let me know if you don't want to work on it.

**Comment by lucyleeow** [comment]
That's fine. We can do it in the sortkey, maybe adding a `parent` ? I think we only ever go 1 folder deep.

**Comment by lucyleeow** [comment]
Sure I'm okay with it the previous status, but it should be documented so people know what we're sorting on (and that it's not just the filename). And tested if possible...

**Comment by lucyleeow** [comment]
Actually, the filepath input arg to the directive would be okay? Since they have to give the filepath relative to `conf.py` so subfolder structure would be retained?

**Comment by lucyleeow** [comment]
Oh yeah good point.

I realise now why I thought it should be the filename, the example we give is along the lines of `x.startswith('plot')`. That would be trickier with the full path, they would have split the path but that's unavoidable if we want the subdir in there.

**Comment by lucyleeow** [comment]
> Ehh, Matplotlib does some wonky nested nonsense so I'd rather be as general as possible.

I meant that I think sphinx gallery only looks 1 level down for subdirs, though matplotlib may do funny things to get around this?

File: sphinx_gallery/directives.py
Comment (lucyleeow) [inline sphinx_gallery/directives.py:142]: @story645 I stumbled on this error accidentally but I don't fully follow what situation this is meant to catch?

I managed to contrive a case where I added a file to a path not in `examples_dir` to a minigallery directive. In that case there were no matches and the error message didn't really make sense.

I assume it is meant to catch cases when `dirs>1`? So maybe a sub-gallery directory is named the same as a examples gallery?

Comment (lucyleeow) [reply sphinx_gallery/directives.py:142]: A more informative error message would be great. There is no test for this but depending on what this is mean to catch, maybe it would be too difficult to write a test.

Comment (lucyleeow) [reply sphinx_gallery/directives.py:142]: cc @larsoner just in case you remember...? :grimacing:

Comment (larsoner) [reply sphinx_gallery/directives.py:142]: No idea :disappointed:

**Pull request merged**
